webknossos-datastore/app/com/scalableminds/webknossos/datastore/explore/ExploreRemoteLayerService.scala

```diff
@@ -106,26 +107,93 @@ class ExploreRemoteLayerService @Inject()(dataVaultService: DataVaultService,
+            .toFox
+          allLayersWithVoxelSizes = layersWithVoxelSizes +: layersWithVoxelSizesInSiblingPaths
+        } yield allLayersWithVoxelSizes.flatten
+
+      case Empty =>
+        for {
+          extendedRemainingPaths <- path.listDirectory().map(dirs => remainingPaths ++ dirs.map((_, searchDepth + 1)))
+          foundLayers <- recursivelyExploreRemoteLayerAtPaths(extendedRemainingPaths,
+                                                              credentialId,
+                                                              explorers,
+                                                              reportMutable)
+        } yield foundLayers
+      case _ =>
+        Fox.successful(List.empty)
+    }
+
+  private def recursivelyExploreRemoteLayerAtPaths(remotePathsWithDepth: List[(VaultPath, Int)],
+                                                   credentialId: Option[String],
+                                                   explorers: List[RemoteLayerExplorer],
+                                                   reportMutable: ListBuffer[String])(
       implicit ec: ExecutionContext): Fox[List[(DataLayerWithMagLocators, VoxelSize)]] =
-    explorers match {
-      case Nil => Fox.empty
-      case currentExplorer :: remainingExplorers =>
-        reportMutable += s"\nTrying to explore $remotePath as ${currentExplorer.name}..."
-        currentExplorer.explore(remotePath, credentialId).futureBox.flatMap {
-          case Full(layersWithVoxelSizes) =>
-            reportMutable += s"Found ${layersWithVoxelSizes.length} ${currentExplorer.name} layers at $remotePath."
-            Fox.successful(layersWithVoxelSizes)
-          case f: Failure =>
-            reportMutable += s"Error when reading $remotePath as ${currentExplorer.name}: ${Fox.failureChainAsString(f)}"
-            exploreRemoteLayersForRemotePath(remotePath, credentialId, reportMutable, remainingExplorers)
-          case Empty =>
-            reportMutable += s"Error when reading $remotePath as ${currentExplorer.name}: Empty"
-            exploreRemoteLayersForRemotePath(remotePath, credentialId, reportMutable, remainingExplorers)
-        }
+    remotePathsWithDepth match {
+      case Nil =>
+        Fox.empty
+      case (path, searchDepth) :: remainingPaths =>
+        if (searchDepth > MAX_RECURSIVE_SEARCH_DEPTH) return Fox.empty
+        implicit val implReportMutable: ListBuffer[String] = reportMutable
+
+        Fox
+          .sequence(explorers.map { explorer =>
+            {
+              explorer
+                .explore(path, credentialId)
+                .futureBox
+                .flatMap {
+                  handleExploreResult(_, explorer, path)
+                }
+                .toFox
+            }
+          })
+          .map(explorationResults => Fox.firstSuccess(explorationResults.map(_.toFox)))
+          .toFox
+          .flatten
+          .futureBox
+          .flatMap(
```

Comment:
Phew, that’s a mouthful. Maybe extracting some of these into functions with annotated return types could help with understanding what’s going on here?

Reply:
Maybe the newest version is easier to read? Each method has something like max 8 lines as a method body